### PR TITLE
CLN: consistent signatures for equals methods

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -1005,19 +1005,20 @@ def _format_space(self) -> str:
...
+    def equals(self, other: object) -> bool:
         """
         Determines if two IntervalIndex objects contain the same elements.
         """
         if self.is_(other):
             return True
 
-        # if we can coerce to an II
-        # then we can compare
+        # if we can coerce to an IntervalIndex then we can compare
```
Comment (simonjayhawkins) [inline pandas/core/indexes/interval.py:1015]: not necessarily in this PR, but is this behaviour consistent with other Index types

```
>>> x = np.linspace(0, 100, num=3)
>>> x
array([  0.,  50., 100.])
>>>
>>> idx = pd.IntervalIndex.from_breaks(x)
>>> idx
IntervalIndex([(0.0, 50.0], (50.0, 100.0]],
              closed='right',
              dtype='interval[float64]')
>>>
>>> arr = idx.values
>>> arr
<IntervalArray>
[(0.0, 50.0], (50.0, 100.0]]
Length: 2, closed: right, dtype: interval[float64]
>>>
>>> idx.equals(arr)
True
>>>
>>> arr.equals(idx)
False
>>>
```

File: pandas/core/arrays/base.py [hunk 2]
```diff
@@ -762,7 +762,8 @@ def equals(self, other: "ExtensionArray") -> bool:
         """
         if not type(self) == type(other):
             return False
-        elif not self.dtype == other.dtype:
+        other = cast(ExtensionArray, other)
+        if not self.dtype == other.dtype:
```
Comment (jreback) [inline pandas/core/arrays/base.py:766]: i guess this is fine, though more consisten to use is_dtype_equal no?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/base.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/base.py:766]: I had to break the if elif to get the cast in. so this is unchanged. but agreed `is_dtype_equal` would be more consistent so will update.

**Review by jreback — APPROVED** [review]
couple of places where codecov thinks we don't have tests, in a followup can remove / test if possible..

**Comment by simonjayhawkins** [comment]
> couple of places where codecov thinks we don't have tests, in a followup can remove / test if possible..

the code added was from static analysis. (mypy is still testing but codecov thinks we don't have tests)

so in this case we don't want to remove.

dynamic type testing has become less necessary now we are using mypy.

**Pull request merged**